File: src/gtm_sourcing_agent/followup_sweep.py
"""Background sweep that auto-sends due outreach follow-ups (Outreach
automation batch). Mirrors task_queue.py's single-in-process-thread
design and its "exactly one server process" assumption (see that
module's docstring) — but runs on a timer instead of consuming a queue,
since there's nothing to enqueue here, just a periodic "what's due right
now" check.

Sends nothing unless db_storage.get_workspace_settings()
["auto_send_followups"] is True — off by default (see WorkspaceSettings'
docstring). This is the one place in the app that emails a candidate
with no recruiter clicking "send" that day, so it stays inert until a
recruiter explicitly opts in.
"""

import logging
import threading
import time
from typing import Any

from . import db_storage, notifications

logger = logging.getLogger(__name__)

SWEEP_INTERVAL_SECONDS = 6 * 60 * 60  # every 6 hours -- a day-granularity cadence doesn't need finer polling

_started = False
_start_lock = threading.Lock()


def run_once() -> list[dict[str, Any]]:
    """Sends every currently-due follow-up if the workspace has opted
    in. Returns the entries actually sent (empty whether that's because
    nothing is due or because auto-send is off, same "no-op when
    unconfigured" convention as notifications.send_email). Exposed
    separately from the sleep loop below so it's directly callable —
    from tests, and from api.py if a manual "run the sweep now" action
    is ever added."""
    settings = db_storage.get_workspace_settings()
    if not settings["auto_send_followups"]:
        return []
    sent: list[dict[str, Any]] = []
    for entry in db_storage.due_followups():
        if not notifications.send_email(
            [entry["email"]], f"Following up: {entry['role_title']}", entry["draft_message"],
        ):
            logger.warning(
                "followup sweep: failed to send stage %s to %s (role %s)",
                entry["followup_stage"], entry["email"], entry["role_id"],
            )
            continue
        db_storage.log_communication(
            entry["role_id"], entry["candidate_id"], channel="email", direction="outbound",
            content=entry["draft_message"], contact_used=entry["email"], logged_by="auto-followup",
            followup_stage=entry["followup_stage"],
        )
        sent.append(entry)
    if sent:
        logger.info("followup sweep: auto-sent %d follow-up(s)", len(sent))
    return sent


def _loop() -> None:
    while True:
        time.sleep(SWEEP_INTERVAL_SECONDS)
        try:
            run_once()
        except Exception:  # the sweep thread must never die over one bad candidate/send
            logger.exception("followup sweep: unhandled error")


def start() -> None:
    global _started
    with _start_lock:
        if _started:
            return
        threading.Thread(target=_loop, name="gtm-followup-sweep", daemon=True).start()
        _started = True

File: backend-node/src/followupSweep.ts
/**
 * Port of followup_sweep.py — background sweep that auto-sends due
 * outreach follow-ups. Runs on a timer instead of consuming a queue,
 * since there's nothing to enqueue here, just a periodic "what's due
 * right now" check.
 *
 * Sends nothing unless getWorkspaceSettings().auto_send_followups is
 * true — off by default. This is the one place in the app that emails
 * a candidate with no recruiter clicking "send" that day, so it stays
 * inert until a recruiter explicitly opts in.
 */
import * as storage from "./db/storage.js";
import { sendEmail } from "./notifications.js";

const SWEEP_INTERVAL_MS = 6 * 60 * 60 * 1000; // every 6 hours

let started = false;

export async function runOnce(): Promise<Record<string, any>[]> {
  const settings = await storage.getWorkspaceSettings();
  if (!settings.auto_send_followups) return [];

  const sent: Record<string, any>[] = [];
  for (const entry of await storage.dueFollowups()) {
    const ok = await sendEmail([entry.email], `Following up: ${entry.role_title}`, entry.draft_message);
    if (!ok) {
      console.warn(
        `followup sweep: failed to send stage ${entry.followup_stage} to ${entry.email} (role ${entry.role_id})`
      );
      continue;
    }
    await storage.logCommunication(entry.role_id, entry.candidate_id, {
      channel: "email", direction: "outbound", content: entry.draft_message,
      contactUsed: entry.email, loggedBy: "auto-followup", followupStage: entry.followup_stage,
    });
    sent.push(entry);
  }
  if (sent.length) console.log(`followup sweep: auto-sent ${sent.length} follow-up(s)`);
  return sent;
}

async function loop(): Promise<void> {
  await new Promise((resolve) => {
    const timer = setTimeout(resolve, SWEEP_INTERVAL_MS);
    timer.unref(); // daemon-like: never keeps the process (or a test run) alive
  });
  try {
    await runOnce();
  } catch (err) {
    console.error("followup sweep: unhandled error", err);
  }
  void loop();
}

export function start(): void {
  if (started) return;
  started = true;
  void loop();
}
